### PR TITLE
Emacs

### PR DESCRIPTION
**Pull request by asmeurer**
This only works on Linux and OS X, and on OS X is built with --without-x (terminal emacs only). This recipe was only tested for the previous version so hopefully it still works.  I did have issues in the past on one of the variants of Linux (I don't remember if it was 64 or 32), so hopefully that doesn't come back either. Maybe someone with one of those variants can install it and see if it actually works. 

TODO:

- [x] `configure` not found
- [ ] `emacs --kill` doesn't work on CircleCI because there is no tty
- [x] dbus is building twice (OS X)
- [x] see if removing `--without-x` works on OS X
- [ ] actually test the packages on a real machine (is that possible to do before merging?)

### CODE REVIEW

**Comment by asmeurer** [comment]
> Failed to even lint the recipe (might be a conda-smithy bug) 😢

I remember that's code for something but I don't remember what.

File: recipes/emacs/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  version: 24.5
+
+source:
+  fn: emacs-24.5.tar.xz
+  url: http://ftp.gnu.org/gnu/emacs/emacs-24.5.tar.xz
+  sha1: 9d65d74506628cec19483204454aee25de5616e6
+
+build:
+  number: 0
+  skip:  # [win]
```
Comment (msarahan) [inline recipes/emacs/meta.yaml]: I think you need a True value here

File: recipes/emacs/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,52 @@
+# Note: This recipe currently only works on OS X and Linux. Furthermore, on OS
+# X, it is built with --without-x (terminal emacs only). Feel free to lift
+# either of these restrictions if you are able to get it to build.
+
+package:
+  name: emacs
+  version: 24.5
```
Comment (msarahan) [inline recipes/emacs/meta.yaml]: Please template the version for easier maintenance

File: recipes/emacs/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    - emacs --kill
+
+about:
+  home: http://www.gnu.org/software/emacs/
+  license: GPL
+  summary: "GNU Emacs is an extensible, customizable text editor—and more."
+
+extra:
+  recipe-maintainers:
+    - asmeurer
```
Comment (msarahan) [inline recipes/emacs/meta.yaml]: Happy to help maintain, if you want.

File: recipes/emacs/build.sh [hunk 4]
```diff
@@ -0,0 +1,7 @@
+if [ "$(uname)" == "Darwin" ]; then
+    ./configure  --prefix=$PREFIX --without-x
+else
+    ./configure  --prefix=$PREFIX --x-includes=$PREFIX/include --x-libraries=$PREFIX/lib
+fi
+
+make -j4  && make -j4  install
```
Comment (msarahan) [inline recipes/emacs/build.sh]: Please don't hard-code the parallel build.  We're working on a better way (allowing pass-through of the CPU_COUNT variable to override autodetection), but it's not ready yet.

Comment (asmeurer) [inline recipes/emacs/build.sh]: So should I just build non-parallel for now? Do the CI VMs even have multiple CPUs?

Comment (msarahan) [inline recipes/emacs/build.sh]: Yeah, for now that's best.  I think Circle CI and AppVeyor offer 2 CPUs, but not Travis.

**Comment by asmeurer** [comment]
I guess CircleCI doesn't have an X toolkit installed. What should I do about that?

**Comment by asmeurer** [comment]
Why is Travis building dbus multiple times?

**Comment by asmeurer** [comment]
And again it can't find `configure`. The recipe builds (or at least gets that far) for me locally. Is it stripping the execute permissions for some reason?

File: recipes/emacs/build.sh [hunk 5]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+if [ "$(uname)" == "Darwin" ]; then
+  OPTS="--without-x"
+else
+  OPTS="--x-includes=$PREFIX/include --x-libraries=$PREFIX/lib"
+fi
+
+bash configure  --prefix=$PREFIX $OPTS
```
Comment (ocefpaf) [inline recipes/emacs/build.sh]: Calling `bash` did not work on `OS X`. One alternative is to `chmod +x configure` or try `sh configure`.

The weird thing is that the source is already... Not sure what is going on.

**Comment by asmeurer** [comment]
So I put an `ls` call at the [top of the build.sh](https://travis-ci.org/conda-forge/staged-recipes/jobs/148701181#L10759):

```
+ ls -al
total 8
drwxr-xr-x   5 travis  staff   170 Jul 31 17:33 .
drwxr-xr-x   9 travis  staff   306 Jul 31 17:32 ..
-rwxrw-rw-   1 travis  staff   316 Jul 31 17:33 conda_build.sh
drwxr-xr-x  54 travis  staff  1836 Jul 31 17:32 dbus-1.8.8
drwxr-xr-x  32 travis  staff  1088 Apr  8  2015 emacs-24.5
```

**Comment by asmeurer** [comment]
Looks like a conda build bug (or conda-build-all, I don't know how that works).

**Comment by asmeurer** [comment]
Just saw https://github.com/conda/conda-build/pull/1175. That should hopefully fix the configure issue.

**Comment by jakirkham** [comment]
Restarted the CIs.

**Comment by asmeurer** [comment]
But was that actually released?

**Comment by asmeurer** [comment]
If so, it wasn't fixed.

**Comment by jakirkham** [comment]
It's true GitHub has been pretty confusing about what shows up in a tag and what doesn't. Not to mention this change was place in `master`. However, there was a `1.21.11` build released today. So if it was backported, it may be in there. It's not clear to me if it was backported so I have [asked]( https://github.com/conda/conda-build/pull/1175#issuecomment-238053079 ).

File: recipes/emacs/build.sh (hunk 5 above)
Comment (jakirkham) [inline recipes/emacs/build.sh]: Could we please try `chmod +x configure` and `./configure`? If that doesn't work, I would be extremely surprised, but maybe it doesn't.

**Comment by jakirkham** [comment]
The [release]( https://github.com/conda/conda-build/releases/tag/1.21.11 ) says it should be in there. So, if that is the cause of the issues and the release log is correct, then it doesn't fix the problem. Though that call to `configure` looks pretty strange to me. Maybe we could try the simple change suggested and see if it doesn't fix it.

**Comment by asmeurer** [comment]
My ls call is still there. It's in the wrong directory.

**Comment by jakirkham** [comment]
Ah, sorry, I'm coming into this conversation late and missed that. So there is a `conda-build-all` issue ( https://github.com/SciTools/conda-build-all/issues/52 ) as well. A fix is in `master`, but has not yet been released.

**Comment by msarahan** [comment]
Please ping me if the conda-build-all fix does not work.  I am pretty sure that I backported @pelson's conda build PR 1175 correctly, but there's always the possibility of an error.

**Comment by jakirkham** [comment]
We could also break this up in pieces, @asmeurer. If you submit a PR for `dbus`, it shouldn't run into this issue and we can get that in. Then we can remove `dbus` from this PR and get `emacs` in too.

**Comment by asmeurer** [comment]
I'm not in a rush.

**Comment by jakirkham** [comment]
Ok, cool. Well then let's wait for a `conda-build-all` release. This should make a good test case for this problem.

**Comment by jakirkham** [comment]
`conda-build-all` version `0.13.2` was released yesterday. Restarted the CIs to see if they now pass.

**Comment by asmeurer** [comment]
Looks like that fixed the configure issue. The other issues (listed at the top) are still valid. Any thoughts on them?

**Comment by asmeurer** [comment]
OK, building the GUI emacs didn't just work on OS X, so I'm not going to bother with it. If someone else wants to try to make it work once this is a feedstock, I'll happy to accept it. For now, I don't personally need it, and I don't feel like trying to figure it out.

**Comment by asmeurer** [comment]
Any idea why dbus is building twice?

**Comment by jakirkham** [comment]
Just noticed `dbus` got merged in PR ( https://github.com/conda-forge/staged-recipes/pull/1361 ). Probably best to remove the copy from here. Would you like to add yourself as a maintainer of [`dbus`]( https://github.com/conda-forge/dbus-feedstock )?

**Comment by asmeurer** [comment]
If other people are maintaining dbus I'm not particularly interested. 

It's probably worth opening an issue for the dbus build issue upstream to one of the conda-forge issue trackers. It looks like a bug in something (conda-build-all?).

**Comment by asmeurer** [comment]
Went ahead and updated this to emacs 25.1.

**Comment by asmeurer** [comment]
OK, I didn't hear any suggestions on the CircleCI issue, so I've just skipped the test there. If anyone knows how to install this package before merging, it would be nice to test it. Otherwise, just merge it (either way, I don't have Linux to test on, so someone else will need to do so).

**Comment by jakirkham** [comment]
We have a Docker container we use for Linux, which is pretty [easy to install]( https://docs.docker.com/machine/install-machine/ ). The image can be gotten by running `docker pull condaforge/linux-anvil`.

**Comment by jakirkham** [comment]
> actually test the packages on a real machine (is that possible to do before merging?)

Also this is a great question. Am working on [this]( https://github.com/conda-forge/staged-recipes/pull/1508 ), which can be used locally. This could be used for some local testing with the Docker container if you would like. Though it needs some tweaks to work on CircleCI.

Wrote up this [proposal]( https://conda-forge.hackpad.com/Staged-Releases-DZNKZdgiMbF ) too about generally providing build artifacts earlier for testing before releasing. Would appreciate your thoughts and feedback on this. It will later be proposed as a CFEP (Conda-Forge Enhancement Proposal).

**Comment by asmeurer** [comment]
So what do you suggest doing for this, specifically for Linux? The `emacs --kill` test makes me pretty confident that this works on OS X, but for Linux, we only test `emacs --help`. Are you suggesting that I pull the docker image, build it locally, and test it in the image?

**Comment by jakirkham** [comment]
I'm suggesting it may be a good way to start working on the problem if you are interested. It also provides you a way to build and test locally. As these are the only unchecked items on your list, maybe you wish to purse that route.

Ultimately being able to run `emacs --kill` on Linux requires that we stop running docker builds with I/O redirection and starting running them with TTY enabled. There are many advantages of running with TTY and testing on Linux is seriously hampered without it (`emacs` is not alone in this respect). IMHO this would be a good switch to make. Though I don't know the history on why we selected to do I/O redirection instead.

This all being said, I think you have a decent working build of `emacs` and we should get out there. Ultimately having some users try it will be a better determination of what it needs.

File: recipes/emacs/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+test:
+  commands:
+    - emacs --help
+
+    # This test fails on CircleCI because there is no tty.
+    - emacs --kill  # [not linux]
+
+about:
+  home: http://www.gnu.org/software/emacs/
+  license: GPL-3.0
```
Comment (jakirkham) [inline recipes/emacs/meta.yaml]: Could we please add `license_file: ` with some sort of license?

Comment (asmeurer) [inline recipes/emacs/meta.yaml]: Done

**Comment by jakirkham** [comment]
Thanks @asmeurer.

**Comment by jakirkham** [comment]
Ah noticed this has the `COPYING` file in the recipe. That's not necessary as `license_file` grabs it from the source not the recipe. Could you please remove that in the feedstock?

**Comment by asmeurer** [comment]
What the hell happened to the OS X package? Can you delete it (or remove it from the `main` label)? 

https://anaconda.org/conda-forge/emacs/files
https://travis-ci.org/conda-forge/emacs-feedstock

**Comment by asmeurer** [comment]
It must have run the system `emacs` in the test. But why didn't it build correctly? https://github.com/conda-forge/emacs-feedstock/pull/1

Anyway, I guess it's always a good idea to use an absolute `$PREFIX` path in the tests.

**Comment by asmeurer** [comment]
Did you want to create an upstream issue for the dbus building bug?

**Comment by jakirkham** [comment]
Sorry to be dense, but what `dbus` building bug? Are you referring to the issue with building more than one package at once in a staged-recipe PR? I believe that got fixed awhile ago, but if you do encounter it again please let us know. FWIW that would have been a [`conda-build-all`]( https://github.com/SciTools/conda-build-all ) bug.

**Comment by asmeurer** [comment]
Before dbus was merged separately, it was part of this PR, and was building twice (but only on OS X).

**Pull request merged**